zabbix_smartmontools/__init__.py

```diff
@@ -1,150 +1,204 @@
+                        break   # break from header check
 
-        sender.append('%s smartctl.info[%s,revision] "%s"' % (host, dR, revisionResult))
+                if duplicate == True:
+                    continue   # continue to the next disk. other sender and json data is discarded
 
-    complianceRe = re.search(r'^Compliance:\s+(.+)$', p, re.M | re.I)
-    if complianceRe:
-        sender.append('%s smartctl.info[%s,compliance] "%s"' % (host, dR, complianceRe.group(1)))
+                driveHeaders.append((serialCurrent, headerCurrent))   # add header info for current disk, only if it's not duplicate
 
-    manufacturedRe = re.search(r'^Manufactured in week (\d+) of year (\d+)', p, re.M | re.I)
-    if manufacturedRe:
-        sender.append('%s smartctl.info[%s,manufacturedYear] "%s"' % (host, dR, manufacturedRe.group(2)))
+        senderData.extend(getSmart_Out[1])
+    return senderData
 
-    # SAS values
-    loadUnloadRe = re.search(r'^Accumulated load-unload cycles:\s+(\d+)', p, re.M | re.I)
-    if loadUnloadRe:
-        sender.append('%s smartctl.value[%s,loadUnload] "%s"' % (host, dR, loadUnloadRe.group(1)))
- 
-        loadUnloadMaxRe = re.search(r'^Specified load-unload count over device lifetime:\s+(\d+)', p, re.M | re.I)
-        if loadUnloadMaxRe:
-            sender.append('%s smartctl.value[%s,loadUnloadMax] "%s"' % (host, dR, loadUnloadMaxRe.group(1)))
 
-    startStopRe = re.search(r'^Accumulated start-stop cycles:\s+(\d+)|^Current start stop count:\s+(\d+)', p, re.M | re.I)
-    if startStopRe:
-        if startStopRe.group(1):
-            startStopResult = startStopRe.group(1)
-        elif startStopRe.group(2):
-            startStopResult = startStopRe.group(2)
+def getDiscoveryData(config, host, diskList):
+    """ Return the data that Zabbix expects for device discovery """
+    reported_serials = set({})
+    jsonData = []
+    for (device, protocol) in diskList:
+        if config['mode'] == 'serial' or config['skipDuplicates']:
+            serial = getSerial(device)
         else:
-            startStopResult = 'UNKNOWN'
-
-        sender.append('%s smartctl.value[%s,startStop] "%s"' % (host, dR, startStopResult))
-
-        startStopMaxRe = re.search(r'^Recommended maximum start stop count:\s+(\d+)|^Specified cycle count over device lifetime:\s+(\d+)', p, re.M | re.I)
-        if startStopMaxRe:
-            if startStopMaxRe.group(1):
-                startStopMaxResult = startStopMaxRe.group(1)
-            elif startStopMaxRe.group(2):
-                startStopMaxResult = startStopMaxRe.group(2)
+            serial = None
+        if config['mode'] == 'serial' and serial:
+            identifier = serial
+        else:
+            identifier = device
+        jsonData.append({'{#DDRIVESTATUS}': identifier})
+        if not serial in reported_serials or not config['skipDuplicates']:
+            jsonData.append({"{#DISKID}": identifier})
+            jsonData.append({"{#DISKIDSAS}": identifier})
```

Comment:
It's a duplicate line (429).
Besides, SAS json should only be discovered when actual SAS disk is present (see the template).

Reply:
Good catch.  But it's not entirely duplicative, because the code at line 429 only runs during GetSmart.  It doesn't run during discovery.  I need to fix the code so that the json data isn't altered at all during getSmart.  Unfortunately, that means I'll have to guess at discovery time which features are supported by which drives.

Reply:
I really don't know how to do that unless by introduction of second (trapper) lld.

Reply:
It will have to involve some guesswork.  Would you be ok with me doing that in a separate PR?

Reply:
Sure.